### ISSUE TITLE
Add email notices for plan visibility changes

### ISSUE THREAD

**Issue by stephaniesimms** [DMPRoadmap/roadmap#732]
Add the 3 new email notifications below (all 3 have same Email title), triggered automatically when a user changes plan visibility on Share page. User can also select to deactivate these auto email notices on Notification Preferences page under Edit Profile.

Email title: DMP Visibility Changed: [PLAN TITLE]

Hello [USER NAME]

The plan [“PLAN TITLE”] had its visibility changed to private.

Visibility definitions:
- Private: restricted to me and my collaborators
- Organisation: anyone at my organisation can view
- Public: anyone can view on the web

If you have questions pertaining to this action, please visit the My Dashboard page in [TOOL NAME]. 

All the best
The [TOOL NAME] team

You may change your notification preferences on your profile page. Please do not reply to this email. If you have any questions or need help, please contact us at [HELPDESK EMAIL] or visit [CONTACT US URL]

Hello [USER NAME]

The plan [“PLAN TITLE”] has had its visibility changed to institutional.

Visibility definitions:
- Private: restricted to me and my collaborators
- Organisation: anyone at my organisation can view
- Public: anyone can view on the web

If you have questions pertaining to this action, please visit the My Dashboard page in [TOOL NAME]. 

All the best
The [TOOL NAME] team

You may change your notification preferences on your profile page. Please do not reply to this email. If you have any questions or need help, please contact us at [HELPDESK EMAIL] or visit [CONTACT US URL]

Hello [USER NAME]

The plan [“PLAN TITLE”] has had its visibility changed to public.

Visibility definitions:
- Private: restricted to me and my collaborators
- Organisation: anyone at my organisation can view
- Public: anyone can view on the web

If you have questions pertaining to this action, please visit the My Dashboard page in [TOOL NAME]. 

All the best
The [TOOL NAME] team

You may change your notification preferences on your profile page. Please do not reply to this email. If you have any questions or need help, please contact us at [HELPDESK EMAIL] or visit [CONTACT US URL]

**Comment by briri** [comment]
This will rely on #760

**Comment by jollopre** [comment]
The mail for visibility changes in plan can be seen at <HOSTNAME>/rails/mailers/user_mailer/plan_visibility

**Comment by stephaniesimms** [comment]
I created a new plan and tried to test visibility changes to trigger email notices on the Share page. Changing my plan visibility to "Org" or "Public" produced an error (screenshot 1) and did not send an email notice @jollopre . 

When I navigated back to My Dashboard, my plan visibility was updated correctly to "Org (Institutional)." screenshot 2

![screen shot 2017-11-13 at 11 51 00 am](https://user-images.githubusercontent.com/12703638/32746323-273574e0-c86a-11e7-9a12-d66b511ad8eb.png)

![screen shot 2017-11-13 at 11 51 58 am](https://user-images.githubusercontent.com/12703638/32746332-2b62fdf8-c86a-11e7-8662-5f82ae0ea14d.png)

**Comment by briri** [comment]
seems to be something wrong with the mailer. here's the log
```
App 23060 stdout: F, [2017-11-13T13:01:42.137632 #23060] FATAL -- : 
App 23060 stdout: ActionView::Template::Error (No route matches {:action=>"index"}):
App 23060 stdout:     31:     <%= _('The %{tool_name} team') %{ :tool_name => tool_name } %>
App 23060 stdout:     32:   </p>
App 23060 stdout:     33:   <p>
App 23060 stdout:     34:     <%=  _('You may change your notification preferences on your profile page.') %>&nbsp;<%= _('Please do not reply to this email.') %>&nbsp;<%= raw(_('If you have any questions or need help, please contact us at %{helpdesk_email} or visit %{contact_us_url}') %{ :helpdesk_email => mail_to(helpdesk_email, helpdesk_email, subject: email_subject), :contact_us_url => link_to(contact_us_url, contact_us_url) }) %>
App 23060 stdout:     35:   </p>
App 23060 stdout:     36: <% end %>
App 23060 stdout:   app/views/user_mailer/plan_visibility.html.erb:34:in `block in _app_views_user_mailer_plan_visibility_html_erb__4555542852065977788_69926481745580'
App 23060 stdout:   app/views/user_mailer/plan_visibility.html.erb:10:in `_app_views_user_mailer_plan_visibility_html_erb__4555542852065977788_69926481745580'
App 23060 stdout:   app/mailers/user_mailer.rb:76:in `block in plan_visibility'
App 23060 stdout:   app/mailers/user_mailer.rb:75:in `plan_visibility'
App 23060 stdout:   app/controllers/plans_controller.rb:317:in `visibility'
```

**Comment by jollopre** [comment]
@stephaniesimms, this error is not development related. At a first glance, the error log facilitated from @briri let me think that perhaps you are missing (tool_name, helpdesk_email or contact_us_url) placeholders. If a placeholder is missing, the string interpolation fails and the view won't get rendered throwing an error.

**Comment by briri** [comment]
We added helpdesk_email in yesterday afternoon related to another ticket. Will check this today to see if its still failing

**Comment by stephaniesimms** [comment]
@briri - can you look into this and ping me to re-test? i just tried again and I get the same error message when I change the visibility setting. the system does save the change. but it doesn't trigger an email notice.

**Comment by briri** [comment]
contact_us_url is not in the branding.yml file nor should it have to be if the installation wants to use the application's built in contact us page. This line overrides Rails' built-in url helper: `contact_us_url = Rails.configuration.branding[:organisation][:contact_us_url]`

I updated this line in the other mailers to use Rails' built-in helper if an override is not defined in the yaml file. Will submit a PR to correct this one.

**Comment by stephaniesimms** [comment]
@jollopre - I can change plan visibility on the Share page now and receive appropriate messages at the top of the page (no more error msgs). but I'm not receiving any email notices. to clarify, when I change the visibility to Private, Org, or Public, an email about that change should be sent to all plan Owners and Co-owners per the Notification preferences setting (screenshot below).

![screen shot 2017-11-20 at 12 48 38 pm](https://user-images.githubusercontent.com/12703638/33040989-aa9e6454-cdf1-11e7-9a32-38b8e0270f56.png)

**Comment by jollopre** [comment]
@stephaniesimms, issue #811 is only handling the condition to send an email given the prefs. I think the visibility email has been setup to send email to just the owner, so I will have to look into this issue again.

**Comment by stephaniesimms** [comment]
all the emails are coming through to all collaborators now. closing issue.

**Comment by jollopre** [comment]
@stephaniesimms, I fixed that as part of #811 :)
